Basic data types documentation for nodejs/abi-stable-node#280.

jschlight opened the pull request:
This is my initial hack at documentation for Basic Types. I'm open to all suggestions and criticisms to improve this document. Please let me know how it can be improved with additional content or format changes.

mhdawson commented on doc/N-API Basic Types.md:
````diff
@@ -0,0 +1,373 @@
+Value is a the base class upon which other JavaScipt values like Number, Boolean, String, and Object are based. 
+
+```
+Value();
````
For each of the methods I would add a heading. In this example:

### Default Constructor

and we might want to add the `Added in: ` which is used in the standard Node doc as well.

We may want to include the Returns line, if if it is just `Returns void` (or whatever we think is the best way to say returns nothing)

See https://nodejs.org/dist/latest-v8.x/docs/api/n-api.html#n_api_napi_create_reference in the N-API doc as the example I looked at to come up with these suggestions.

mhdawson commented on doc/N-API Basic Types.md:
```diff
@@ -0,0 +1,373 @@
+# N-API Documentation
```
A general thought is that some intro at the top that introduces the basic types and how they are used might be useful before the section with the details.  We can add that later though so what you have here would be good to land without that.

mhdawson replied:
We might want to put a big "WORK IN PROGRESS, NOT YET COMPLETE" at the top just to set peoples expectations.

mhdawson commented:
Overall a good start just a few suggestions.   I also think we can incrementally land changes so if you like my suggestions we could integrate those (and any others made in the mean time), land what you have so far and then continue to add content in follow on PRs.

jschlight commented:
I've split the earlier single large md file into four separate files that fit into the new documentation scaffolding structure. I've implemented Michael's suggestions with the exception of the `Added in:` marker which will require further research.

mhdawson commented on doc/callbackinfo.md:
````diff
@@ -0,0 +1,99 @@
+size_t Length() const;
+```
+
+Returns the number of arguemnts passed in the Callabckinfo object.
````
Callbackinfo -> CallbackInfo  (capital I)

mhdawson replied:
Just noticed arguemnts -> arguments

mhdawson replied:
arguemnts -> arguments

mhdawson commented on doc/callbackinfo.md:
````diff
@@ -0,0 +1,99 @@
+Value NewTarget() const;
+```
+
+Returns the `new.target` value of the constructor call. If the CallbackInfo is not a constructor call, a call to `IsEmpty()` on the returned value returns true.
````
This reads a bit strangely (to my ears).  Maybe:

If the function that was invoked (and for which the CallbackInfo was passed) is not a constructor call....

There is also similar wording in a few other places.

mhdawson commented on doc/callbackinfo.md:
```diff
@@ -0,0 +1,99 @@
+
+Returns the JavaScript `this` value for the call
+
+### Data
```
Would be good to provide a bit more details on what/how Data is used but that can be added in a follow on PR.

jschlight replied:
I'm not clear how Data is used. I'm assuming that it's a place the C++ programmer can store general information with the C++ object, but I'm not sure. If so, I can add that to the documentation.

mhdawson replied:
I think we can add the info as a follow on PR we just need to record somewhere that we should come back and do that.  We should generate a test that uses it which would provide an example to base the doc on.

mhdawson commented on doc/env.md:
```diff
@@ -0,0 +1,65 @@
+
+The opaque data structure containing the environment in which the request is being run.
+
+The Env object is usually created and passed by the calling system.
```
Would it be better to say 

"is usually created and passed by the Node.js runtime or node-addon-api infrastructure"

The same comment also applies other places were we say "calling system"

mhdawson commented on doc/external.md:
````diff
@@ -0,0 +1,68 @@
+```cpp
+template <typename T>
+static External New(napi_env env,
+			T* data,
````
The parameters don't line up here.

mhdawson commented on doc/external.md:
````diff
@@ -0,0 +1,68 @@
+```cpp
+template <typename T>
+static External New(napi_env env,
+			T* data,
````
parameters don't line up here

mhdawson commented on doc/external.md:
```diff
@@ -0,0 +1,68 @@
+Returns a pointer to the arbitrary C++ data held by the External object.
+
+### Not documented here
+
```
Is this just because the doc will be added in future PR ?  Or do you have some questions that need to be answered before you can add the doc for these ?

jschlight replied:
I wanted the reviewers to know that they're not documented. I'm pretty sure there's no need to document them. If so, I'll remove the note.

mhdawson replied:
I'm thinking best to remove.  It confused me so it might confuse others as well.

jschlight commented:
Michael, thank you for your comments. I've implemented your recommendations, have pushed updated files, and added a couple of comments above.

jschlight commented:
Thanks, Michael. I've made another push to my branch.

mhdawson commented:
Landed as 7ecab460df01413f523fb7216d4b03db1fd2b8c9
